neurolib/models/multimodel/builder/base/network.py

```diff
@@ -0,0 +1,717 @@
+
+    @property
+    def state_variable_names(self):
+        state_var_names = sum([mass.state_variable_names for mass in self], [])
+        # if state variables are not unique per node, append mass label to them
+        if len(state_var_names) != len(set(state_var_names)):
+            state_var_names = [
+                f"{state_var}_{mass.mass_type}" for mass in self for state_var in mass.state_variable_names
+            ]
+        return [state_var_names]
+
+    @property
+    def max_delay(self):
+        return 0.0
+
+    def get_nested_parameters(self):
+        """
+        Return nested dictionary with parameters from all masses within this
+        node.
+
+        :return: nested dictionary with all parameters
+        :rtype: dict
+        """
+        assert self.initialised
+        node_key = f"{NODE_NAME_STR}_{self.index}"
+        nested_dict = {node_key: {}}
+        for mass in self:
+            mass_key = f"{MASS_NAME_STR}_{mass.index}"
+            nested_dict[node_key][mass_key] = mass.parameters
+        return nested_dict
+
+    def init_node(self, **kwargs):
+        """
+        Initialise node and all the masses within.
+
+        :kwargs: optional keyword arguments to init_mass
+        """
+        # we need to have the index already assigned
+        assert self.index is not None
+        # initialise possible sync variables
+        self.sync_symbols = {
+            f"{symbol}_{self.index}": se.Symbol(f"{symbol}_{self.index}") for symbol in self.sync_variables
+        }
+        for mass in self:
+            mass.init_mass(**kwargs)
+        assert all(mass.initialised for mass in self)
+        self.initialised = True
+
+    def update_parameters(self, parameters_dict):
+        """
+        Update parameters of the node, i.e. recursively update all the masses.
+
+        :param parameters_dict: new parameters for this node, same format as
+            `get_nested_parameters`, i.e. nested dict
+        :type parameters_dict: dict
+        """
+        for mass_key, mass_params in parameters_dict.items():
+            if MASS_NAME_STR in mass_key:
+                mass_index = int(mass_key.split("_")[-1])
+                assert mass_index == self.masses[mass_index].index
+                self.masses[mass_index].update_parameters(mass_params)
+            else:
+                logging.warning(f"Not sure what to do with {mass_key}...")
+
+    @staticmethod
+    def _strip_index(symbol_name):
+        """
+        Strip index value from the symbol name.
+        """
+        return "_".join(symbol_name.split("_")[:-1])
+
+    def _callbacks(self):
+        """
+        Gather callbacks from all masses.
+        """
+        return sum([mass._callbacks() for mass in self.masses], [])
+
+    def _numba_callbacks(self):
+        """
+        Gather callbacks from all masses.
+        """
+        return sum([mass._numba_callbacks() for mass in self.masses], [])
+
+    def _sync(self):
```

Comment:
`_sync()` function is the main coupling power... in the node's sync one should define how are masses connected into one node, similar in the network case (below) the sync method defines how are nodes connected into a network

Reply:
side-note the coupling itself (using `_sync()` method) is implemented as `jitcdde`'s helpers... i.e. this should be a list of tuples `(variable_name, <dynamics definition in sympy>)` where you can imagine `variable_name` being `network_x` (as in the case of Hopf) and here the coupling of `x` variables within the Hopf network model is done. the definition for x-coupling is simply a connectivity matrix * inputs (with correct delays as per delay matrix)...

in the end, when actually integrating, all the `sync` tuples from each node and the `sync` from network (plus any tentative helpers from masses, e.g. for AdEx) are concatenated into one huge list and this is passed into the `jitcdde` for integration... for `numba` backend, these are directly substituted into the symbolic definition.